### PR TITLE
Module cleanup + Potion "durability" + Documentation

### PR DESCRIPTION
**Pull request by xrtariq2594**
### Contains

* Moved potion prefabs from Alchemy module to this one.
* Cleanup performed on the potion prefabs and related components to remove unnecessary stuff.
* Potions can now be turned back into their original bottle prefabs as long as they have enough durability remaining (#1).
* Added documentation to all classes.

### How to Test

* Spawn potions using the console, and drink them. If no potion bottle is returned, then it's a success.

* To test the other scenario, modify one of the potion prefabs to include the following code:
```
    "Durability": {
        "durability": 9,
        "maxDurability": 9
    },
```
Then, spawn that potion and drink it. It should return an empty GlassBottle.

* Proofread the documentation.

### Issues
None so far.

### Credits

@AWebb3701 and @Cervator for advice.

### CODE REVIEW

File: src/main/java/org/terasology/potions/HerbEffect.java [hunk 1]
```diff
@@ -17,8 +17,31 @@
...
+ * Interface for the application of various herb effects over a period of time.
+ */
 public interface HerbEffect {
+    /**
+     * Apply an herb effect on the entity for the given magnitude and duration.
+     *
+     * @param instigator    The instigator who is applying this effect on the entity. It can be a herb, potion, etc.
+     * @param entity        The entity who the herb effect is being applied on.
+     * @param magnitude     The magnitude of the effect.
+     * @param duration      The duration of the effect.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/HerbEffect.java]: What is the unit of `duration`? Imho this should be stated in the javadoc, e.g., `"The duration of the effect in seconds"`

File: src/main/java/org/terasology/potions/HerbEffect.java [hunk 2]
```diff
@@ -17,8 +17,31 @@
...
+     * Apply an herb effect with a specific ID on the entity for the given magnitude and duration. This specific ID
+     * can be used to differentiate effects, such as the different types of damage over time effects. It's up to the
+     * individual HerbEffect implementations to decide how to use this.
+     *
+     * @param instigator    The instigator who is applying this effect on the entity. It can be a herb, potion, etc.
+     * @param entity        The entity who the herb effect is being applied on.
+     * @param id            This particular HerbEffect's ID. Can be used to differentiate different effects under the
+     *                      same family.
+     * @param magnitude     The magnitude of the effect.
+     * @param duration      The duration of the effect.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/HerbEffect.java]: missing unit spec (2)

File: src/main/java/org/terasology/potions/PotionCommonEffects.java [hunk 3]
```diff
@@ -15,7 +15,10 @@
  */
 package org.terasology.potions;
 
-// Class containing a list of constant Strings used for potion-related effect type checks.
+/**
+ * This class contains a list of constant Strings used for potion-related effect type checks. Most are self-explanatory
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/PotionCommonEffects.java]: I would call the "constant Strings" just constants. Are these the effect IDs?

File: src/main/java/org/terasology/potions/component/PotionEffect.java [hunk 4]
```diff
@@ -17,10 +17,15 @@
 
 import org.terasology.reflection.MappedContainer;
 
-// This class or container contains details on a particular effect of a potion.
+/** This class or mapped container contains details on a particular effect of a potion. */
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/component/PotionEffect.java]: Afaik, the mapped container allows to serialize the effect list. I'd just go for "Common class for describing a potion effect" or something like that.

File: src/main/java/org/terasology/potions/effect/AlterationEffectWrapperHerbEffect.java [hunk 5]
```diff
@@ -19,18 +19,46 @@
 import org.terasology.entitySystem.entity.EntityRef;
 import org.terasology.potions.HerbEffect;
 
+/**
+ * This HerbEffect wraps any AlterationEffect so that the latter can function within the herbalism systems and be treated
+ * as a HerbEffect.
+ */
 public class AlterationEffectWrapperHerbEffect implements HerbEffect {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/effect/AlterationEffectWrapperHerbEffect.java]: Neatpicking on naming 😛 what about `AlterationToHerbEffectWrapper`?

File: src/main/java/org/terasology/potions/effect/DoNothingEffect.java [hunk 6]
```diff
@@ -18,11 +18,33 @@
...
+ * in case one or more of the listed effects in prefabs are invalid.
+ */
 public class DoNothingEffect implements HerbEffect {
+    /**
+     * Do nothing when attempting to apply an effect to the given entity.
+     *
+     * @param instigator    The instigator who is applying this effect on the entity. It can be a herb, potion, etc.
+     * @param entity        The entity who the herb effect is being applied on.
+     * @param magnitude     The magnitude of the effect.
+     * @param duration      The duration of the effect.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/effect/DoNothingEffect.java]: duration unit spec (3)

File: src/main/java/org/terasology/potions/effect/DoNothingEffect.java [hunk 7]
```diff
@@ -18,11 +18,33 @@
...
 
+    /**
+     * Do nothing when attempting to apply an effect to the given entity.
+     *
+     * @param instigator    The instigator who is applying this effect on the entity. It can be a herb, potion, etc.
+     * @param entity        The entity who the herb effect is being applied on.
+     * @param id            This particular HerbEffect's ID. Can be used to differentiate different effects under the
+     *                      same family.
+     * @param magnitude     The magnitude of the effect.
+     * @param duration      The duration of the effect.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/effect/DoNothingEffect.java]: as above

File: src/main/java/org/terasology/potions/effect/HealEffect.java [hunk 8]
```diff
@@ -20,14 +20,37 @@
 import org.terasology.logic.health.DoHealEvent;
 import org.terasology.math.TeraMath;
 
+/**
+ * This HerbEffect heals the target entity when applied.
+ */
 public class HealEffect implements HerbEffect {
+    /** The maximum healing allowed. Used more as a a baseline measure to be later multiplied by the magnitude. */
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/effect/HealEffect.java]: should this be renamed then? maybe `targetHeal`, `targetHP`, or `baseHeal`?

File: src/main/java/org/terasology/potions/effect/HealEffect.java [hunk 9]
```diff
@@ -20,14 +20,37 @@
 import org.terasology.logic.health.DoHealEvent;
 import org.terasology.math.TeraMath;
 
+/**
+ * This HerbEffect heals the target entity when applied.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/effect/HealEffect.java]: should mention that healing is applied instantly.

File: src/main/java/org/terasology/potions/events/BeforeDrinkPotionEvent.java [hunk 10]
```diff
@@ -21,119 +21,233 @@
...
-import org.terasology.potions.HerbEffect;
 import org.terasology.potions.component.PotionComponent;
 import org.terasology.potions.component.PotionEffect;
 
+/**
+ * This event is sent to an entity to allow modification and cancellation of potion drinks.
+ * Note that one event will be sent per potion effect, so individual effects can be cancelled.
+ */
 // TODO: Add extends AbstractValueModifiableEvent?
 public class BeforeDrinkPotionEvent implements ConsumableEvent {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/events/BeforeDrinkPotionEvent.java]: if the event is sent for each effect, it should be named `BeforeApplyPotionEffect` imho. `BeforeDrinkPotion` should/could be used to cancel the drinking alltogether (maybe it is not allowed, or some other system added a "potion cooldown")

File: src/main/java/org/terasology/potions/events/BeforeDrinkPotionEvent.java [hunk 11]
```diff
@@ -173,11 +294,20 @@ public double getDurationResultValue() {
         return result;
     }
 
+    /**
+     * Informs whether or not this event has been consumed.
+     *
+     * @return True if the the event has been consumed, false otherwise.
+     */
     @Override
     public boolean isConsumed() {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/events/BeforeDrinkPotionEvent.java]: you may checkout `{@inheritDoc}` and use cases. this case seems like the javadoc should be same in the parent class (the consumable event). If you don't write javadoc at all, it is automatically copied from the override method.

File: src/main/java/org/terasology/potions/events/DrinkPotionEvent.java [hunk 12]
```diff
@@ -20,34 +20,76 @@
 import org.terasology.entitySystem.event.Event;
 import org.terasology.potions.component.PotionComponent;
 
+/**
+ * This event is sent to an entity to indicate that it'll be drinking a given potion.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/events/DrinkPotionEvent.java]: is this event treating the potion as one or is it sent for each effect, as the `BeforeDrink` event?

File: src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java [hunk 13]
```diff
@@ -43,23 +49,58 @@
...
+    /**
+     * Check the potion drink event by sending out a consumable event so that other systems can either modify the
+     * effect parameters or cancel (consume) the event outright. Following that, if not consumed, apply the effect to
+     * the instigator.
+     * This calls the {@link #checkDrink(EntityRef, EntityRef, PotionComponent, HerbEffect, PotionEffect, String)}}
+     * method but leaves the id field empty.
+     *
+     * @param instigator    The entity who is trying to consume this potion.
+     * @param item          A reference to the potion item entity which has the potion effect.
+     * @param p             The PotionComponent of the potion item.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java]: although well documented, please remove the abbreviated variable names and change them to something more clear

File: src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java [hunk 14]
```diff
@@ -161,13 +213,51 @@ public void onPotionWithoutGenomeConsumed(DrinkPotionEvent event, EntityRef ref)
...
         }
 
+        // Play the potion drink sound.
         audioManager.playSound(Assets.getSound("engine:drink").get(), 1.0f);
+
+        // Get the EntityRef of the potion, and its durability component (if any).
+        EntityRef potion = event.getItem();
+        DurabilityComponent durability = potion.getComponent(DurabilityComponent.class);
+
+        // If the Durability component exists.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java]: some people might argue about the usefulness of such comments... I don't mind them, but in this case it is very clear from the code what is happening

File: src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java [hunk 15]
```diff
@@ -161,13 +213,51 @@ public void onPotionWithoutGenomeConsumed(DrinkPotionEvent event, EntityRef ref)
...
+        if (durability != null) {
+            // If the new durability value will be above 0 following the potion drink, or the bottle has inf durability, proceed.
```
Comment (skaldarnar) [inline src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java]: this one is fine, though ;-)

**Review by skaldarnar — COMMENTED** [review]
Great work, and probably one of the best documented pieces of code in the whole project 😆 
Have a look at the comments, mabye they are helpful at some point. 

Tested out the durability, and it looks to work just right.

File: src/main/java/org/terasology/potions/HerbEffect.java (hunk 1 above)
Comment (xrtariq2594) [reply src/main/java/org/terasology/potions/HerbEffect.java]: Done. And actually the duration is in milliseconds.

File: src/main/java/org/terasology/potions/PotionCommonEffects.java (hunk 3 above)
Comment (xrtariq2594) [reply src/main/java/org/terasology/potions/PotionCommonEffects.java]: Yep, they are basically effect IDs.

File: src/main/java/org/terasology/potions/component/PotionEffect.java (hunk 4 above)
Comment (xrtariq2594) [reply src/main/java/org/terasology/potions/component/PotionEffect.java]: Okay, going with: 

> Common class for describing an effect that may be present in a potion.

File: src/main/java/org/terasology/potions/effect/AlterationEffectWrapperHerbEffect.java (hunk 5 above)
Comment (xrtariq2594) [reply src/main/java/org/terasology/potions/effect/AlterationEffectWrapperHerbEffect.java]: Done! That was an old class from older maintainer (from WoodAndStone in fact).

File: src/main/java/org/terasology/potions/effect/HealEffect.java (hunk 8 above)
Comment (xrtariq2594) [reply src/main/java/org/terasology/potions/effect/HealEffect.java]: I renamed it to your third suggestion.

File: src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java (hunk 13 above)
Comment (xrtariq2594) [reply src/main/java/org/terasology/potions/system/DrinkPotionAuthoritySystem.java]: Done. I un-abbreviated the names.

File: src/main/java/org/terasology/potions/events/DrinkPotionEvent.java (hunk 12 above)
Comment (xrtariq2594) [reply src/main/java/org/terasology/potions/events/DrinkPotionEvent.java]: Treated the Potion as one.

DrinkPotionEvent sends the entire potion.
BeforeDrinkPotionEvent (now titled BeforeApplyPotionEffectEvent) works on one potion effect at a time.

File: src/main/java/org/terasology/potions/events/BeforeDrinkPotionEvent.java (hunk 10 above)
Comment (xrtariq2594) [reply src/main/java/org/terasology/potions/events/BeforeDrinkPotionEvent.java]: Done.

I'll add BeforeDrinkPotionEvent (the new one to completely cancel the potion drink event) in the next feature release.

File: src/main/java/org/terasology/potions/events/BeforeDrinkPotionEvent.java (hunk 11 above)
Comment (xrtariq2594) [reply src/main/java/org/terasology/potions/events/BeforeDrinkPotionEvent.java]: It is done.

**Comment by xrtariq2594** [comment]
Overall, I thank you for the suggestions @skaldarnar . I applied to them to the module, and I'll be pushing them tomorrow.

**Comment by skaldarnar** [comment]
Changes look good! What keeps you from merging this? Anything outstanding to be tested?

**Comment by xrtariq2594** [comment]
Nothing. I was just waiting for your confirmation before merging @skaldarnar.

Now that I have it, I'll merge this in now.

**Pull request merged**